lib/src/main/kotlin/com/swmansion/starknet/provider/rpc/JsonRpcProvider.kt

```diff
@@ -68,6 +66,28 @@ class JsonRpcProvider(
         return HttpRequest(payload, buildJsonHttpDeserializer(responseSerializer, deserializationJson), httpService)
     }
 
+    fun <T> combineBatchRequest(
+        calls: List<Request<T>>,
+    ): BatchHttpRequest<T> {
+        val httpRequests = calls.map { it as HttpRequest<T> }
+        val updatedJsonObjects = httpRequests.mapIndexed { index, request ->
+            val jsonMap = Json.parseToJsonElement(request.payload.body.toString()).jsonObject.toMutableMap()
+            jsonMap["id"] = JsonPrimitive(index)
+            JsonObject(jsonMap)
+        }
+
+        val jsonArray = JsonArray(updatedJsonObjects)
+        val deserializer = httpRequests.first().deserializer
```

Comment:
can you explain this to me, since I'm not sure I follow - why do we take .first() here?
can `calls` be ever empty? what happens then?

Reply:
It seems to me every request should have it's own `deserializer`

Reply:
Here we need to determine the deserializer for our response. Since `httpRequests` have the same generic type, we retrieve deserializer from the first one. I've added prevention from retrieving first element's deserializer when `calls` is empty.

Reply:
I've updated this part, now every request has it's own deserializer.